### PR TITLE
feat(import/export): #MAG-43 Import export "mes données"

### PR DESCRIPTION
**Pull request by Alex-Delaveau**
## Describe your changes
added import export support throught module mes données
## Checklist tests
Add some data in Magneto, then go to the "Mes Données" module, export data from Magneto, and try to import the file that the export provided. Check in the Magneto module if the data has been imported.
## Issue ticket number and link
[MAG-43](https://jira.support-ent.fr/browse/MAG-43)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Date;
+
+public class ImportExportHelper {
+
+    public static JsonObject setModifiedMagneto(JsonObject doc, Date date)
+    {
+        try
+        {
+            // Check that the dates are in string format
+            String m = doc.getString("modified");
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java:17]: magic string

File: src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        return doc;
+    }
+
+
+    public static JsonObject revertExportChanges(JsonObject document, String importPrefix) {
+        String title = DocumentHelper.getTitle(document);
+        String name = DocumentHelper.getName(document);
+        String headline = DocumentHelper.getAppProperty(document, "headline");
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java:46]: magic string

File: src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        if (title != null && title.startsWith(importPrefix)) {
+            DocumentHelper.setTitle(document, title.substring(importPrefix.length()));
+        }
+
+        if (name != null && name.startsWith(importPrefix)) {
+            DocumentHelper.setName(document, name.substring(importPrefix.length()));
+        }
+
+        if (headline != null && headline.startsWith(importPrefix)) {
+            DocumentHelper.setAppProperty(document, "headline", headline.substring(importPrefix.length()));
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/helper/ImportExportHelper.java:56]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+    protected static final Logger log = LoggerFactory.getLogger(DefaultBoardService.class);
+
+    public MagnetoRepositoryEvents(Vertx vertx) {
+        super(vertx);
+        this.collectionNameToImportPrefixMap.put(CollectionsConstant.FOLDER_COLLECTION, "folder_");
+        this.collectionNameToImportPrefixMap.put(CollectionsConstant.BOARD_COLLECTION, "board_");
+        this.collectionNameToImportPrefixMap.put(CollectionsConstant.CARD_COLLECTION, "card_");
+        this.collectionNameToImportPrefixMap.put(CollectionsConstant.SECTION_COLLECTION, "section_");
+    }
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:31]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+            boards.stream().map(JsonObject.class::cast).forEach(board -> {
+                boardIds.add(board.getString("_id"));
+            });
+
+            JsonObject query = new JsonObject()
+                    .put("boardId", new JsonObject().put("$in", boardIds));
+
+            mongo.find("magneto.sections", query, res -> {
+                JsonArray results = res.body().getJsonArray("results");
+                if ("ok".equals(res.body().getString("status")) && results != null) {
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:75]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+                    promise.fail("Could not process query" + query.encodePrettily());
+                }
+            });
+        });
+        return promise.future();
+    }
+
+
+    private void addResultsWithPrefix(JsonArray results, JsonArray collectionResults, String prefix) {
+        collectionResults.forEach(item -> ((JsonObject) item).put("title", prefix + ((JsonObject) item).getString("title")));
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:87]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        );
+        QueryBuilder findByAuthorOrShared = exportSharedResources == false ? findByAuthor : QueryBuilder.start().or(findByAuthor.get(), findByShared.get());
+        JsonObject query;
+
+        if (resourcesIds == null)
+            query = MongoQueryBuilder.build(findByAuthorOrShared);
+        else {
+            QueryBuilder limitToResources = findByAuthorOrShared.and(
+                    QueryBuilder.start("_id").in(resourcesIds).get()
+            );
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:120]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+        }
+
+        //Check for duplicates
+        QueryBuilder lookForExisting = QueryBuilder.start("_id").in(oldIds);
+
+        mongo.find(collection, MongoQueryBuilder.build(lookForExisting), searchMsg -> {
+            JsonObject body = searchMsg.body();
+
+            if (body.getString("status").equals("ok")) {
+                JsonArray foundDocs = body.getJsonArray("results");
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:283]: magic string

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+                                readFilePromise.complete();
+                            });
+                        });
+                    }
+                    CompositeFuture.all(readFileFutures).onComplete(res -> {
+                        if (res.succeeded()) {
+                            promise.complete(collectionsContent);
+                        } else {
+                            promise.fail(res.cause());
+                        }
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:372]: log.info à ajouter

File: src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+     */
+    private Future<JsonArray> findInSections(Future<JsonArray> boardsFuture) {
+        Promise<JsonArray> promise = Promise.promise();
+        List<String> boardIds = new ArrayList<>();
+        boardsFuture.onSuccess(boards -> {
+            boards.stream().map(JsonObject.class::cast).forEach(board -> boardIds.add(board.getString(Field._ID)));
+
+            JsonObject query = new JsonObject()
+                    .put(Field.BOARDID, new JsonObject().put("$in", boardIds));
+
```
Comment (lapeyronnie) [inline src/main/java/fr/cgi/magneto/service/impl/MagnetoRepositoryEvents.java:66]: magic string

**Review by BastienBenard — APPROVED** [review]

**Review by Sofianebernoussi — APPROVED** [review]

**Pull request merged**